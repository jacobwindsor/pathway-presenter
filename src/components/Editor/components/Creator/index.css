.creator-wrapper {
    position: relative;
    height: 100%;
    width: 100%;
    overflow: hidden;
}

.creator-wrapper .left-section,.right-section {
    height: 100%;
}

.creator-wrapper .left-section {
    width: 30%;
    float: left;
}

.creator-wrapper .right-section {
    width: 70%;
    margin-left: calc(30%);
    height: calc(100% - 56px);
}

.creator-wrapper .right-section .diagram-comp-wrapper {
    height: 80%;
    width: 100%;
}

.creator-wrapper .add-slide-button {
    position: absolute;
    bottom: 10%;
    transform: translateY(50%);
    right: 3rem;
}