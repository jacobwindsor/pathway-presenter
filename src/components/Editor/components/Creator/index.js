import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PreviewPanel from './components/PreviewPanel';
import EditorPanel from './components/EditorPanel';
import Paper from 'material-ui/Paper';
import Title from '../../../Title';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import Divider from 'material-ui/Divider';
import Diagram from '../../../Diagram';
import ContentAdd from 'material-ui/svg-icons/content/add';
import { findIndex, cloneDeep } from 'lodash';
import Toolbar from './components/Toolbar';
import './index.css'

class Creator extends Component {
    constructor(props) {
        super(props);
        this.state = {
            activeSlideIndex: 0,
            selectedEntity: null,
            presentation: cloneDeep(props.presentation),
        }
    }

    handlePreviewClick = (slideNumber) => {
        this.setState({
            // slideNumber is not a 0th index
            activeSlideIndex: slideNumber - 1,
            selectedEntity: null,
        });
    };

    handleEntityClick = (entity) => {
        // For now we only allow entities with text (nodes)
        if (! entity.textContent ) return;
        this.setState({selectedEntity: entity})
    };

    handleSlideUpdate = (slide) => {
        const { activeSlideIndex } = this.state;
        this.setState(state => {
            let newSlides = state.presentation.slides.slice();
            newSlides[activeSlideIndex] = slide;
            return {
                presentation: Object.assign({}, state.presentation, {
                    slides: newSlides
                }),
                selectedEntity: null,
            }
        })
    };

    handleSlideRemove = slideIndex => {
        this.setState(state => {
            const copy = cloneDeep(state.presentation);
            copy.slides.splice(slideIndex, 1);
            return {
                presentation: copy,
                selectedEntity: null
            }
        })
    };

    onSlideAdd = () => {
        this.setState(state => {
            return {
                presentation: Object.assign({}, state.presentation, {
                    slides: state.presentation.slides.concat({
                        targets: [],
                        title: null
                    }),
                }),
                // Since added one slide, new index is the length
                activeSlideIndex: state.presentation.slides.length,
                selectedEntity: null,
            }
        })
    };

    handleSave = () => {
        const { presentation } = this.state;
        const { handleSave } = this.props;
        handleSave(presentation);
    };

    render() {
        const { activeSlideIndex, selectedEntity, presentation } = this.state;
        const slide = presentation.slides[activeSlideIndex];

        return (
            <div className="creator-wrapper">
                <Toolbar handleSave={this.handleSave} />
                <div className="left-section">
                    <EditorPanel
                        slide={slide}
                        slideIndex={activeSlideIndex}
                        activeEntity={selectedEntity}
                        onUpdate={this.handleSlideUpdate}
                    />
                </div>
                <div className="right-section">
                    <div className="previewer">
                        <div className="slide-wrapper">
                            <div className="slide">
                                <Paper className="content" zDepth={2}>
                                    { slide.title ? <Title title={slide.title} /> : null }
                                    <Diagram
                                        wpId={presentation.wpId}
                                        version={presentation.version}
                                        detailPanelEnabled={false}
                                        onEntityClick={this.handleEntityClick}
                                        slide={slide}
                                        showPanZoomControls={true} />
                                </Paper>
                            </div>
                        </div>
                    </div>
                    <Paper className="footer">
                        <PreviewPanel
                            slides={presentation.slides}
                            wpId={presentation.wpId}
                            version={presentation.version}
                            onClick={this.handlePreviewClick}
                            handleSlideRemove={this.handleSlideRemove}
                            width={'calc(100% - 10rem)'}
                            height="100%"
                        />
                        <FloatingActionButton
                            className="add-slide-button"
                            onTouchTap={this.onSlideAdd}
                            tooltip="Add a new slide"
                        >
                            <ContentAdd />
                        </FloatingActionButton>
                    </Paper>
                </div>
            </div>
        )
    }
}

Creator.propTypes = {
    presentation: PropTypes.object.isRequired,
    handleSave: PropTypes.func.isRequired,
};

export default Creator;