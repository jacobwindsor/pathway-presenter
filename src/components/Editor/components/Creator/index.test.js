import React from 'react';
import { shallow } from 'enzyme';
import Creator from './index';
import toJson from 'enzyme-to-json';

describe('The Creator. Interface for actually creating the presentations with all controls', () => {
  it('should render correctly', () => {
    const presentation = {
      title: 'My title',
      wpId: 'WP4',
      authorName: 'Ronald McDonald',
      version: 0,
      slides: [
        {
          title: 'My slide',
          targets: []
        }
      ]
    };

    const comp = shallow(
      <Creator
        presentation={presentation}
        handleSave={() => null}
        handleDelete={() => null}
      />
    );
    expect(toJson(comp)).toMatchSnapshot();
  });
});
