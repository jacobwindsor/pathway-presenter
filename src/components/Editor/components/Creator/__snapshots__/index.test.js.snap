// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Creator. Interface for actually creating the presentations with all controls should render correctly 1`] = `
<div
  className="creator-wrapper"
>
  <EditorToolbar
    authorName="Ronald McDonald"
    handlePresentClick={[Function]}
    handleSave={[Function]}
    handleSettingsClick={[Function]}
    title="My title"
  />
  <SettingsDialog
    authorName="Ronald McDonald"
    handleClose={[Function]}
    handleDelete={[Function]}
    handleSave={[Function]}
    isOpen={false}
    title="My title"
    version={0}
    wpId="WP4"
  />
  <EmptyComps />
  <span>
    <div
      className="left-section"
    >
      <EditorPanel
        activeEntity={null}
        handleCancelTarget={[Function]}
        handleTargetChipClick={[Function]}
        hasSlideChanged={true}
        onUpdate={[Function]}
        slide={
          Object {
            "targets": Array [],
            "title": "My slide",
          }
        }
        slideIndex={0}
      />
    </div>
    <div
      className="right-section"
    >
      <div
        className="previewer"
      >
        <div
          className="slide-wrapper"
        >
          <_class2
            diagramLocked={false}
            handleEntityClick={[Function]}
            handleLock={[Function]}
            handlePanZoomChanged={[Function]}
            handleUnlock={[Function]}
            slide={
              Object {
                "targets": Array [],
                "title": "My slide",
              }
            }
            version={0}
            wpId="WP4"
          />
        </div>
      </div>
      <Paper
        circle={false}
        className="footer"
        rounded={true}
        transitionEnabled={true}
        zDepth={1}
      >
        <PreviewPanel
          activeSlideIndex={0}
          handleUpdate={[Function]}
          height="100%"
          onClick={[Function]}
          slides={
            Array [
              Object {
                "targets": Array [],
                "title": "My slide",
              },
            ]
          }
          version={0}
          width="calc(100% - 10rem)"
          wpId="WP4"
        />
        <FloatingActionButton
          className="add-slide-button"
          disabled={false}
          mini={false}
          onTouchTap={[Function]}
          secondary={false}
          zDepth={2}
        >
          <ContentAdd />
        </FloatingActionButton>
      </Paper>
    </div>
  </span>
</div>
`;
